src/dowel/csv_output.py

```diff
@@ -42,20 +42,42 @@ def record(self, data, prefix=''):
                 self._writer.writeheader()
 
             if to_csv.keys() != self._fieldnames:
-                self._warn('Inconsistent TabularInput keys detected. '
-                           'CsvOutput keys: {}. '
-                           'TabularInput keys: {}. '
-                           'Did you change key sets after your first '
-                           'logger.log(TabularInput)?'.format(
-                               set(self._fieldnames), set(to_csv.keys())))
-
-            self._writer.writerow(to_csv)
+                self._fieldnames |= to_csv.keys()
+                if set(to_csv.keys()).issubset(self._fieldnames):
+                    for key in self._fieldnames:
+                        if key not in to_csv.keys():
+                            to_csv[key] = ''
+                    self._writer.writerow(to_csv)
+                else:
+                    self.add_key()
+            else:
+                self._writer.writerow(to_csv)
 
             for k in to_csv.keys():
                 data.mark(k)
         else:
             raise ValueError('Unacceptable type.')
 
+    def add_key(self, to_csv):
+        """
+        add key or keys from to_csv if they are not present in current log
+        :param to_csv: data to be added to csv
+        """
+        rewrite_csv = list()
+        reader = csv.DictReader(self._log_file)
+        fieldnames = reader.fieldnames
+        for row in reader:
+            for key in to_csv.keys():
+                if key not in fieldnames:
+                    row[key] = ''
```

Comment:
This logic can actually be replaced with setting the `restval` option to the `DictWriter`.